delta/sdk-testkit/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/testkit/RealmsDummy.scala

```diff
@@ -64,93 +44,54 @@ final class RealmsDummy private (
 
   override def fetchAt(label: Label, rev: Long): IO[RealmRejection.RevisionNotFound, Option[RealmResource]] =
-    stateAt(label, rev).map(_.flatMap(_.toResource))
+    journal
+      .stateAt(label, rev, Initial, Realms.next, RealmRejection.RevisionNotFound.apply)
+      .map(_.flatMap(_.toResource))
 
   override def list(pagination: FromPagination, params: RealmSearchParams): UIO[UnscoredSearchResults[RealmResource]] =
-    cache.get.map { resources =>
-      val filtered = resources.values.filter(params.matches).toVector.sortBy(_.createdAt)
-      UnscoredSearchResults(
-        filtered.length.toLong,
-        filtered.map(UnscoredResultEntry(_)).slice(pagination.from, pagination.from + pagination.size)
-      )
-    }
-
-  override def events(offset: Offset = NoOffset): Stream[Task, Envelope[RealmEvent]] =
-    DummyHelpers.eventsFromJournal(journal.get, offset, maxStreamSize)
-
-  override def currentEvents(offset: Offset): Stream[Task, Envelope[RealmEvent]] =
-    DummyHelpers.currentEventsFromJournal(journal.get, offset, maxStreamSize)
-
-  private def setToCache(resource: RealmResource): UIO[RealmResource]         =
-    cache.update(_ + (resource.id -> resource)).as(resource)
-
-  private def deprecateFromCache(resource: RealmResource): UIO[RealmResource] =
-    cache.update(_.updatedWith(resource.id)(_.as(resource))).as(resource)
-
-  private def currentState(label: Label): UIO[Option[RealmState]] =
-    journalMap.map { labelsEvents =>
-      labelsEvents
-        .get(label)
-        .map(_.foldLeft[RealmState](Initial) { (s, e) => Realms.next(s, e.event) })
-    }
-
-  private def stateAt(label: Label, rev: Long): IO[RevisionNotFound, Option[RealmState]] =
-    journalMap.flatMap { labelsEvents =>
-      labelsEvents.get(label).traverse { events =>
-        if (events.size < rev)
-          IO.raiseError(RevisionNotFound(rev, events.size.toLong))
-        else
-          events
-            .foldLeft[RealmState](Initial) {
-              case (state, envelope) if envelope.event.rev <= rev => Realms.next(state, envelope.event)
-              case (state, _)                                     => state
-            }
-            .pure[UIO]
-      }
-    }
+    cache.list(pagination, params)
 
   private def eval(cmd: RealmCommand): IO[RealmRejection, RealmResource] =
-    semaphore.withPermit {
-      for {
-        lbEvents <- journal.get
-        state    <- currentState(cmd.label).map(_.getOrElse(Initial))
-        event    <- Realms.evaluate(resolveWellKnown, cache.get.map(_.values.toSet))(state, cmd)
-        _        <- journal.set(lbEvents :+ makeEnvelope(event, s"$entityType-${event.label}", offsetMax))
-        res      <- IO.fromEither(Realms.next(state, event).toResource.toRight(UnexpectedInitialState(cmd.label)))
-      } yield res
-    }
+    for {
+      state <- journal.currentState(cmd.label, Initial, Realms.next).map(_.getOrElse(Initial))
```

Comment:
I think it's important to make these evaluations "thread" safe

Reply:
Understood